File: Results/DER/Translation/translation_output/w_test/codeqwen/codenet/python/java/p02932.java
Here is the Java version of the provided Python code:

```java
import java.util.Scanner;

public class Main {
    static final int MOD = 1000000007;
    static final int U = 600005;
    static int[] fact = new int[U + 1];
    static int[] fact_inv = new int[U + 1];

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int N = scanner.nextInt();
        int M = scanner.nextInt();
        int L = scanner.nextInt();
        int R = scanner.nextInt();

        fact[0] = 1;
        for (int n = 1; n <= U; n++) {
            fact[n] = (fact[n - 1] * n) % MOD;
        }

        fact_inv[U] = pow(fact[U], MOD - 2, MOD);
        for (int n = U; n > 0; n--) {
            fact_inv[n - 1] = (fact_inv[n] * n) % MOD;
        }

        int[] comb = new int[U + 1];
        int U_ = N - M;
        comb[U_] = fact[U_] * fact_inv[U_] % MOD * fact_inv[U_] % MOD;
        for (int i = U_ - 1; i >= 0; i--) {
            comb[i] = (comb[i + 1] * (M + i + 1)) % MOD;
        }

        int[] P = new int[R + 1];
        for (int i = M; i <= R; i++) {
            P[i - M] = comb[i];
        }

        for (int i = M; i < R; i++) {
            P[i + 1] = (P[i + 1] - P[i]) % MOD;
        }

        int[] Q = new int[R + 1];
        for (int i = 0; i <= R; i++) {
            Q[i

File: Results/DER/Translation/translation_output/w_test/codeqwen/codenet/java/python/p02932.py
Here is the Python version of the provided Java code:

```python
import sys
import math
import random

class Modular:
    def __init__(self, m):
        self.m = m

    def valueOf(self, x):
        x %= self.m
        if x < 0:
            x += self.m
        return x

    def mul(self, x, y):
        return self.valueOf(x * y)

    def plus(self, x, y):
        return self.valueOf(x + y)

    def subtract(self, x, y):
        return self.valueOf(x - y)

class BitOperator:
    def bitAt(self, x, i):
        return (x >> i) & 1

    def setBit(self, x, i, v):
        if v:
            x |= 1 << i
        else:
            x &= ~(1 << i)
        return x

    def swapBit(self, x, i, j):
        bi = self.bitAt(x, i)
        bj = self.bitAt(x, j)
        x = self.setBit(x, i, bj == 1)
        x = self.setBit(x, j, bi == 1)
        return x

class Power:
    def __init__(self, modular):
        self.modular = modular

    def pow(self, x, n):
        if n == 0:
            return 1
        r = self.pow(x, n >> 1)
        r = self.modular.valueOf(r * r)
        if (n & 1) == 1:
            r = self.modular.valueOf(r * x)
        return r

    def inverse(self, x):
        return self.pow(x, self.modular.m - 2)

class Log2:
    def ceilLog(self, x):
        return 32 - int(math.log2(x - 1))

    def floorLog(self, x):
        return 31 - int(math.log2(x))

class InverseNumber:
    
